xds/internal/xdsclient/tests/ads_stream_flow_control_test.go

```diff
@@ -0,0 +1,623 @@
+//   - Starts a management server and configures a listener resource on it.
+//   - Creates an xDS client to the above management server, starts a couple of
+//     listener watchers for the above resource, and verifies that the update
+//     reaches these watchers.
+//   - These watchers don't invoke the onDone callback until explicitly
+//     triggered by the test. This allows the test to verify that the next
+//     Recv() call on the ADS stream does not happen until both watchers have
+//     completely processed the update, i.e invoke the onDone callback.
+//   - Resource is update on the management server, and the test verifies that
+//     the update reaches the watchers.
+func (s) TestADSFlowControl_ResourceUpdates_SingleResource(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	// Override the ADS stream creation.
+	adsStreamCh := overrideADSStreamCreation(t)
+
+	// Start an xDS management server.
+	mgmtServer := e2e.StartManagementServer(t, e2e.ManagementServerOptions{})
+
+	// Create bootstrap configuration pointing to the above management server.
+	nodeID := uuid.New().String()
+	bc := e2e.DefaultBootstrapContents(t, nodeID, mgmtServer.Address)
+	testutils.CreateBootstrapFileForTesting(t, bc)
+
+	// Create an xDS client with the above bootstrap contents.
+	client := createXDSClient(t, bc)
+
+	// Configure two watchers for the same listener resource.
+	const listenerResourceName = "test-listener-resource"
+	const routeConfigurationName = "test-route-configuration-resource"
+	watcher1 := newBLockingListenerWatcher()
+	cancel1 := xdsresource.WatchListener(client, listenerResourceName, watcher1)
+	defer cancel1()
+	watcher2 := newBLockingListenerWatcher()
+	cancel2 := xdsresource.WatchListener(client, listenerResourceName, watcher2)
+	defer cancel2()
+
+	// Wait for the wrapped ADS stream to be created.
+	var adsStream *wrappedADSStream
+	select {
+	case adsStream = <-adsStreamCh:
+	case <-ctx.Done():
+		t.Fatalf("Timed out waiting for ADS stream to be created")
+	}
+
+	// Configure the listener resource on the management server.
+	resources := e2e.UpdateOptions{
+		NodeID:         nodeID,
+		Listeners:      []*v3listenerpb.Listener{e2e.DefaultClientListener(listenerResourceName, routeConfigurationName)},
+		SkipValidation: true,
+	}
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatalf("Failed to update management server with resources: %v, err: %v", resources, err)
+	}
+
+	// Ensure that there is a read on the stream.
+	select {
+	case <-adsStream.recvCh:
+	case <-ctx.Done():
+		t.Fatalf("Timed out waiting for ADS stream to be read from")
+	}
+
+	// Wait for the update to reach the watchers.
+	select {
+	case <-watcher1.updateCh:
+	case <-ctx.Done():
+		t.Fatalf("Timed out waiting for update to reach watcher 1")
+	}
+	select {
+	case <-watcher2.updateCh:
+	case <-ctx.Done():
+		t.Fatalf("Timed out waiting for update to reach watcher 2")
+	}
```

Comment:
Aren't these receives flaky? The sends have a default and there's no guarantee this testing goroutine gets to these lines before the OnUpdate callbacks send on the channel right?

Reply:
Oh I guess since it's verifying *something* has updated it (but this would pass if numerous updates since next updates would get dropped). Do we care about this/is it important?

Reply:
> Aren't these receives flaky? The sends have a default and there's no guarantee this testing goroutine gets to these lines before the OnUpdate callbacks send on the channel right?

No, they aren't flaky because the channels have a buffer of size `1`. So, even if testing goroutine is not reading on the channels when the update fires, the writes on the channels will go through.

> Oh I guess since it's verifying something has updated it (but this would pass if numerous updates since next updates would get dropped). Do we care about this/is it important?

Our concern is not the number of times an update/error callback is invoked. Instead, it is the about when the next read happens. And if there is a bug in the xDS client which invokes the callbacks multiple times for a single update, that should be caught by other tests.

Here, we ensure that multiple reads on the ADS stream is not happening because we do have a check for that, a few lines below.

Hope this is satisfactory for you.

Reply:
Alright sounds good.